test/functional/wallet_coinselection.py

```diff
@@ -0,0 +1,261 @@
+        # This is more than cost of change and BnB won't find a solution with single input
+        # We either get
+        # 1) 2x bech32 solution (SRD only)
+        # 2) legacy + bech32 solution
+        tx = select_coins_for(0.99994899)
+        # legacy + dropping more for fees is still better in terms of waste than legacy + bech32
+        # TODO: implement optimization and update asserts
+        assert_equal(len(tx['vin']), 2)
+        assert_equal(len(tx['vout']), 2)
+
+        # a solution with change is the best solution
+        # 2x bech32 is still more efficient than bech32-and-legacy, but only could be found with SRD
+        tx = select_coins_for(0.99994220)
+        assert_equal(len(tx['vin']), 2)
+        assert_equal(len(tx['vout']), 2)
+
+    def test_one_big_and_many_small_coins(self):
+        w0 = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+        self.nodes[0].createwallet(wallet_name="test_one_big_and_many_small_coins")
+        w1 = self.nodes[0].get_wallet_rpc("test_one_big_and_many_small_coins")
+
+        # Create a bunch of small coins
+        for i in range(100):
+            w0.sendtoaddress(w1.getnewaddress("", "legacy"), 0.05)
+
+        # We will target a solution with 20 coins that will add up exactly to 1btc.
+        # Add a coin that will cover all the fees and exactly match the target.
+        # TODO: fix that knapsack targets change even if there is no change needed
+        # (148*20+42+31)*350+5000000 = 6061550
+        w0.sendtoaddress(w1.getnewaddress("", "legacy"), 0.0606155)
+        self.generate(self.nodes[0], 6)
+
+        fee_rate = 350
+        dummyaddress = w0.getnewaddress()
+
+        def select_coins(big_coin, target, iterations=100):
+            txid = w0.sendtoaddress(w1.getnewaddress("", "bech32"), big_coin)
+            self.generate(self.nodes[0], 6)
+
+            fee_total = 0
+            max_fee = 0
+            for _ in range(iterations):
+                result = w1.fundrawtransaction(w1.createrawtransaction([], {dummyaddress: target}), {"fee_rate": fee_rate})
+                fee_total += result['fee']
+                max_fee = max(max_fee, result['fee'])
+
+            # lock big coin
+            w1.lockunspent(False, [tx for tx in w1.listunspent() if tx['txid'] == txid])
+            return fee_total, max_fee
+
+        # in all the scenarios below the best solution would be to take one big coin and create a change output
+        best_fee = (42 + 68 + 31) * fee_rate / COIN
+
+        # 0) no exact match is possible, the big coin is closest to target + MIN_CHANGE,
+        # knapsack always finds the best solution
+        fee_total, max_fee = select_coins(big_coin=1.01051, target=1.00000001, iterations=100)
+        assert_approx(max_fee, best_fee)
+        assert_approx(fee_total, best_fee * 100)
+
+        # 1) knapsack latches to the exact match solution and can't find the best solution
```

Comment:
We've previously used "exact match" for any changeless solution, so it wasn't immediately obvious to me that it was used in the literal sense here. Perhaps you should clarify that you mean that it's literally the same to the sat.

Reply:
Will update the comment

Reply:
Updated, but I think we should stop using "exact match" in place of changeless solution. Changeless solution is a solution in the range around exact match constrained by the cost of change/discard fee rate.